src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,323 @@
+  primitive_add_char (zmod n) F' :=
+begin
+  let nn := n.to_pnat hn.1,
+  haveI : ne_zero ((nn : ℕ) : F') := ⟨h⟩,
+  haveI : ne_zero ((nn : ℕ) : cyclotomic_field nn F') := ne_zero.of_no_zero_smul_divisors F' _ n,
+  exact
+{ n := nn,
+  char := zmod_char n (is_cyclotomic_extension.zeta_pow nn F' _),
+  prim := zmod_char_primitive_of_primitive_root n (is_cyclotomic_extension.zeta_spec nn F' _) }
+end
+
+/-!
+### Existence of a primitive additive character on a finite field
+-/
+
+/-- There is a primitive additive character on the finite field `F` if the characteristic
+of the target is different from that of `F`.
+We obtain it as the composition of the trace from `F` to `zmod p` with a primitive
+additive character on `zmod p`, where `p` is the characteristic of `F`. -/
+noncomputable
+def primitive_char_finite_field (F F': Type) [field F] [fintype F] [field F']
+ (h : ring_char F' ≠ ring_char F) :
+  primitive_add_char F F' :=
+begin
+  let p := ring_char F,
+  haveI hp : fact p.prime := ⟨char_p.char_is_prime F _⟩,
+  haveI hp₁ : fact (0 < p) := ⟨hp.1.pos⟩,
+  have hp₂ : ¬ ring_char F' ∣ p :=
+  begin
+    cases char_p.char_is_prime_or_zero F' (ring_char F') with hq hq,
+    { exact mt (nat.prime.dvd_iff_eq hp.1 (nat.prime.ne_one hq)).mp h.symm, },
+    { rw [hq],
+      exact λ hf, nat.prime.ne_zero hp.1 (zero_dvd_iff.mp hf), },
+  end,
+  let ψ := primitive_zmod_char p F' (ne_zero_iff.mp (ne_zero.of_not_dvd F' hp₂)),
+  let ψ' := ψ.char.comp (add_monoid_hom.to_multiplicative
+                          (algebra.trace (zmod p) F).to_add_monoid_hom),
+  have hψ' : is_nontrivial ψ' :=
+  begin
+    obtain ⟨a, ha⟩ := finite_field.trace_to_zmod_nondegenerate F one_ne_zero,
+    rw one_mul at ha,
+    exact
+      ⟨a, λ hf, ha $ (primitive_zmod_char_eq_one_iff p ψ.prim $ algebra.trace (zmod p) F a).mp hf⟩,
+  end,
+  exact
+{ n := ψ.n,
+  char := ψ',
+  prim := is_primitive_of_is_nontrivial ψ' hψ' },
+end
+
+/-!
+### The sum of all character values
+-/
+
+open_locale big_operators
+
+variables [fintype R]
+
+/-- The sum over the values of a nontrivial additive character vanishes if the target ring
+is a domain. -/
+lemma sum_eq_zero_of_is_nontrivial [is_domain R'] {ψ : add_char R R'} (hψ : is_nontrivial ψ) :
+  ∑ a, ψ a = 0 :=
+begin
+  rcases hψ with ⟨b, hb⟩,
+  have h₁ : ∑ (a : R), ψ (of_add (b + a)) = ∑ (a : R), ψ a :=
+    fintype.sum_bijective _ (add_group.add_left_bijective b) _ _ (λ x, rfl),
+  simp_rw [of_add_add, ψ.map_mul] at h₁,
+  have h₂ : ∑ (a : R), ψ (of_add a) = finset.univ.sum ⇑ψ := rfl,
+  rw [← finset.mul_sum, h₂] at h₁,
+  exact eq_zero_of_mul_eq_self_left hb h₁,
+end
+
+/-- The sum over the values of the trivial additive character is the cardinality of the source. -/
+lemma sum_eq_card_of_is_trivial {ψ : add_char R R'} (hψ : ¬ is_nontrivial ψ) :
+  ∑ a, ψ a = fintype.card R :=
+begin
+  simp only [is_nontrivial] at hψ,
+  push_neg at hψ,
+  simp only [hψ, finset.sum_const, nat.smul_one_eq_coe],
+  refl,
+end
```

Comment:
With my changes, this proof breaks because we're summing over `multiplicative R`. Here's a proposed fix if you decide to go this way:
```lean
lemma sum_eq_card_of_is_trivial {ψ : add_char R R'} (hψ : ¬ is_nontrivial ψ) :
  ∑ a, ψ (of_add a) = fintype.card R :=
begin
  simp only [is_nontrivial] at hψ,
  push_neg at hψ,
  simp only [hψ, finset.sum_const, nat.smul_one_eq_coe],
  refl,
end

lemma sum_eq_card_of_is_trivial' {ψ : add_char R R'} (hψ : ¬ is_nontrivial ψ) :
  ∑ a, ψ a = fintype.card R :=
sum_eq_card_of_is_trivial hψ
```
Note that I abuse defeq in the proof of the second lemma but the statements are all type correct in the strong sense.

Reply:
I was working on getting the statements correct, but I guess I better wait until you are done...

Reply:
I assume the same applies to `sum_eq_zero_of_is_nontrivial`. I'll make the necessary changes.

Reply:
No, wait -- the statement was OK, and the proof is still valid, so no change necesary.